core/state_transition.go

```diff
@@ -236,8 +232,11 @@ func (st *StateTransition) buyGas() error {
 		balanceCheck = balanceCheck.Mul(balanceCheck, st.gasFeeCap)
 		balanceCheck.Add(balanceCheck, st.value)
 		if st.evm.ChainConfig().Scroll.FeeVaultEnabled() {
-			// always add l1fee, because all tx are L2-to-L1 ATM
-			balanceCheck.Add(balanceCheck, st.l1Fee)
+			// should be fine to add st.l1DataFee even without `L1MessageTx` check, since L1MessageTx will come with 0 l1DataFee,
+			// but double check to make sure
+			if !st.msg.IsL1MessageTx() {
```

Comment:
We short-circuit in `preCheck()` for `L1MessageTx` so we won't call `buyGas()` for it. But I don't mind leaving some extra checks here.